src/engine/output-results.ts

```diff
@@ -25,64 +24,199 @@ interface NodeData {
  * Engine function to take in the ELM patient results and output data relating to the patient's pathway
  * @param pathway - the entire pathway
  * @param patientData - the data on the patient from a CQL execution. Note this is a single patient not the entire patientResults object
- * @return returns PathwayResults describing
+ * @return PathwayResults describing
  */
 export function pathwayData(
   pathway: Pathway,
   patientData: PatientData,
   resources: DomainResource[]
 ): PathwayResults {
   const startNode = 'Start';
-  const patientDocumentation: { [key: string]: Documentation } = {};
-
-  let currentNodes = [startNode];
-  let nodeData = nextNode(pathway, patientData, startNode, resources);
-  while (nodeData !== null) {
-    if (nodeData.documentation === null) break;
-    const documentation = nodeData.documentation;
-    documentation.onPath = true;
-    (documentation as DocumentationResource).resource = retrieveResource(
-      nodeData.documentation,
-      resources
-    );
-    patientDocumentation[nodeData.documentation.node] = { ...documentation };
-    if (nodeData.nextNodes.length === 0) break;
-    else if (nodeData.nextNodes.length === 1) {
-      currentNodes = nodeData.nextNodes;
-      nodeData = nextNode(pathway, patientData, nodeData.nextNodes[0], resources);
-    } else {
-      // There are multiple transitions
-      // Check if any of them have been done
-      currentNodes = [];
-      const completedNodes: string[] = [];
-      for (const nodeKey of nodeData.nextNodes) {
-        const documentReference = retrieveNote(pathway.nodes[nodeKey].label, resources);
-        if (!documentReference && (!patientData[nodeKey] || !patientData[nodeKey].length)) {
-          currentNodes.push(nodeKey);
+  let currentNodes: string[] = [];
+  let nextNodes: string[] = [startNode];
+  const documentation: { [key: string]: Documentation } = {};
+
+  // think of this function like a proof by induction:
+  // - basis: we start at the start node
+  // - recursive: we know the current node(s) we are on, find the next ones
+  // if there are next ones, repeat the process with those, until we don't find any next steps
+  // then our "current node(s)" at the end, are the current step shown to the user
+
```

Comment:
I put a debugger right here and I noticed the `patientData` object (returned from executing the elm` is sometimes an empty list instead of `false`

Reply:
The whole `patientData` object is a list? Or the fields within the object are sometimes a list? Originally the CQL in the pathway would always produce tuples/psuedo-resources, which if none was returned then we would get an empty list. The shift to boolean-valued criteria actually only changes the branch transition criteria, the CQL for action nodes should still return that pseudo-resource, so if an action is not present on the record it will return an empty list in the patient results. (If you break at the debugger again after accepting an action, it should come back with something in the list)

Reply:
Ahhh that makes sense